### DISCUSSION TITLE
How do I create a copy of my add-in's current workbook?

### DISCUSSION THREAD

**Discussion by PabloXV** [exceljs/exceljs#2224] [Q&A] [answered]
I am building an Excel Add-in and would like to create a new workbook that is a copy of the currently open workbook.

I am able to create an empty workbook using Excel.createWorkbook(), I am able to read the currently open file as plaintext using Office.context.document.getFileAsync(), and I am even able to create a duplicate of an existing file using Excel.createWorkbook(base64) where base64 is the output of an online converter like [this one](https://base64.guru/converter/encode/file).

However, I am unable to connect these pieces together. In particular, I've found that Office.context.document.getFileAsync() and its documentation don't clearly explain how to get the necessary base64 format of the file. As I said, I can load the file as plaintext or a byte array, but not base64. I believe that once I can figure this part out, I can use the base64 to create a duplicate of the currently open file.

In the docs, there is a [section](https://learn.microsoft.com/en-us/office/dev/add-ins/excel/excel-add-ins-workbooks#create-a-workbook) about creating a copy of an existing workbook. The docs go on to say that:

> You can get your add-in's current workbook as a base64-encoded string by using [file slicing](https://learn.microsoft.com/en-us/javascript/api/office/office.document?view=common-js-preview#office-office-document-getfileasync-member(1)).

The docs claim that we can get the current workbook as a base64-encoded string, but the function mentioned requires a FileType which only has 3 options: Text, Compressed Byte Array, and PDF. None of these are base64 encoded. Do I have to manually convert these myself? The documentation is a bit lacking in this area. Any help is appreciated!

**Comment by PabloXV** [comment, marked as answer]
Well, I finally broke down and pasted this entire question into ChatGPT. It worked. The key was to combine all of the byte arrays using file slicing (basically just concatenating the slices into one large array) and then throwing it into `Buffer.from(new Uint8Array(fileContent)).toString('base64');`.


```
  Office.context.document.getFileAsync(Office.FileType.Compressed, { sliceSize: 65536 },
    function(result) {
      if (result.status == "succeeded") {
        const file = result.value;
        const sliceCount = file.sliceCount;
        let slicesReceived = 0;
        const fileContent = [];

        function readFile(sliceIndex) {
          file.getSliceAsync(sliceIndex, (sliceResult) => {
            try {
              if (sliceResult.status === "succeeded") {
                const sliceData = sliceResult.value.data;
                const byteArray = new Uint8Array(sliceData);
                for (let i = 0; i < byteArray.length; i++) {
                  fileContent.push(byteArray[i]);
                }
                slicesReceived++;

                if (slicesReceived === sliceCount) {
                  const base64Data = Buffer.from(new Uint8Array(fileContent)).toString('base64');
                  Excel.createWorkbook(base64Data);
                } else {
                  readFile(sliceIndex + 1);
                }
              }
            } catch (err) {
              console.log(err);
            }
          });
        }

        readFile(0);
      } else {
        console.log(result);
      }
    }
  );
```

**Reply by MalikSamiAwan** [reply]
does that work for @PabloXV excel on web
because i tried this my addin is side loaded linked to web and its not working as expected

**Reply by PabloXV** [reply]
I haven't tried it on web, sorry! My use case is mainly for Desktop.
